lib/slots.ts

```diff
@@ -39,21 +43,249 @@ export const NANOSECONDS = 'slot-nanoseconds';
+
+  // Date, YearMonth, MonthDay common slots
+  [DATE_BRAND]: SlotInfo<true, Temporal.PlainDate>;
+  [YEAR_MONTH_BRAND]: SlotInfo<true, Temporal.PlainYearMonth>;
+  [MONTH_DAY_BRAND]: SlotInfo<true, Temporal.PlainMonthDay>;
+
+  // ZonedDateTime
+  [INSTANT]: SlotInfo<Temporal.Instant, Temporal.ZonedDateTime>;
+  [TIME_ZONE]: SlotInfo<Temporal.TimeZoneProtocol, Temporal.ZonedDateTime>;
+
+  // Duration
+  [YEARS]: SlotInfo<number, Temporal.Duration>;
+  [MONTHS]: SlotInfo<number, Temporal.Duration>;
+  [WEEKS]: SlotInfo<number, Temporal.Duration>;
+  [DAYS]: SlotInfo<number, Temporal.Duration>;
+  [HOURS]: SlotInfo<number, Temporal.Duration>;
+  [MINUTES]: SlotInfo<number, Temporal.Duration>;
+  [SECONDS]: SlotInfo<number, Temporal.Duration>;
+  [MILLISECONDS]: SlotInfo<number, Temporal.Duration>;
+  [MICROSECONDS]: SlotInfo<number, Temporal.Duration>;
+  [NANOSECONDS]: SlotInfo<number, Temporal.Duration>;
+
+  // Calendar
+  [CALENDAR_ID]: SlotInfo<BuiltinCalendarId, Temporal.Calendar>;
+}
+
+type TypesWithCalendarUnits =
+  | Temporal.PlainDateTime
+  | Temporal.PlainDate
+  | Temporal.PlainTime
+  | Temporal.PlainYearMonth
+  | Temporal.PlainMonthDay
+  | Temporal.ZonedDateTime;
+
+interface SlotsToTypes {
+  // Instant
+  [EPOCHNANOSECONDS]: Temporal.Instant;
+
+  // TimeZone
+  [TIMEZONE_ID]: Temporal.TimeZone;
+
+  // DateTime, Date, Time, YearMonth, MonthDay
+  [ISO_YEAR]: TypesWithCalendarUnits;
+  [ISO_MONTH]: TypesWithCalendarUnits;
+  [ISO_DAY]: TypesWithCalendarUnits;
+  [ISO_HOUR]: TypesWithCalendarUnits;
+  [ISO_MINUTE]: TypesWithCalendarUnits;
+  [ISO_SECOND]: TypesWithCalendarUnits;
+  [ISO_MILLISECOND]: TypesWithCalendarUnits;
+  [ISO_MICROSECOND]: TypesWithCalendarUnits;
+  [ISO_NANOSECOND]: TypesWithCalendarUnits;
+  [CALENDAR]: TypesWithCalendarUnits;
+
+  // Date, YearMonth, MonthDay common slots
+  [DATE_BRAND]: Temporal.PlainDate;
+  [YEAR_MONTH_BRAND]: Temporal.PlainYearMonth;
+  [MONTH_DAY_BRAND]: Temporal.PlainMonthDay;
+
+  // ZonedDateTime
+  [INSTANT]: Temporal.ZonedDateTime;
+  [TIME_ZONE]: Temporal.ZonedDateTime;
+
+  // Duration
+  [YEARS]: Temporal.Duration;
+  [MONTHS]: Temporal.Duration;
+  [WEEKS]: Temporal.Duration;
+  [DAYS]: Temporal.Duration;
+  [HOURS]: Temporal.Duration;
+  [MINUTES]: Temporal.Duration;
+  [SECONDS]: Temporal.Duration;
+  [MILLISECONDS]: Temporal.Duration;
+  [MICROSECONDS]: Temporal.Duration;
+  [NANOSECONDS]: Temporal.Duration;
+
+  // Calendar
+  [CALENDAR_ID]: Temporal.Calendar;
+}
+
+type SlotKey = keyof SlotsToTypes;
+
 const slots = new WeakMap();
-export function CreateSlots(container) {
+export function CreateSlots(container: AnyTemporalType): void {
   slots.set(container, Object.create(null));
 }
-function GetSlots(container) {
+
+function GetSlots<T extends AnyTemporalType>(container: T) {
   return slots.get(container);
 }
-export function HasSlot(container, ...ids) {
+// TODO: is there a better way than 9 overloads to make HasSlot into a type
```

Comment:
My understanding is that this can be done by:
  - wrapping each of the elements in the variadic args array type into a tuple (`[keyType1, keyType2]` becomes `[[keyType1], [keyType2]]`)
  - convert this into an intersection type (following https://fettblog.eu/typescript-union-to-intersection/) (`[[keyType1], [keyType2]]` becomes `[keyType1]|[keyType2]`)
  - unwrap the tuples  (`[keyType1]|[keyType2]` becomes `keyType1|keyType2`)
  
  Though I can try and sort this out in a separate PR if you don't want to take a shot at that.

Reply:
Yeah, I'd like to land this big PR quickly. If you want to take a shot in a later PR that'd be awesome!